src/Compilers/Core/Portable/Text/ChangedText.cs

```diff
@@ -240,149 +240,187 @@ private static ImmutableArray<TextChangeRange> Merge(IReadOnlyList<ImmutableArra
-                            // delete as much from old change as new change can
-                            // a new change deletion is a reduction in the old change insertion
-                            var oldChangeReduction = Math.Min(oldChange.NewLength, newChange.Span.Length);
-                            AddRange(list, new TextChangeRange(oldChange.Span, oldChange.NewLength - oldChangeReduction));
-                            oldDelta = oldDelta - oldChange.Span.Length + (oldChange.NewLength - oldChangeReduction);
-                            oldIndex++;
-
-                            // deduct the amount removed from oldChange from newChange's deletion span (since its already been applied)
-                            newChange = new TextChangeRange(new TextSpan(oldChange.Span.Start + oldDelta, newChange.Span.Length - oldChangeReduction), newChange.NewLength);
-                            goto nextOldChange;
-                        }
+
+                        // adjust the new change as follows:
+                        // 1. add the old change's deletion to its deletion
+                        // 2. subtract the old change's insertion from its deletion
+                        // 3. add oldDelta to its start position so that its adjusted start position is the same as the oldChange that we removed
+                        //    - note that this can result in the newChange having a negative Start, but this gets adjusted back into a non-negative Start in the next iteration.
+                        var newDeletion = newChange.Span.Length + oldChange.Span.Length - oldChange.NewLength;
+                        newChange = new TextChangeRange(new TextSpan(oldChange.Span.Start + oldDelta, newDeletion, validateStart: false), newChange.NewLength);
+                        if (tryGetNext(oldChanges, ref oldIndex, out oldChange)) continue;
+                        else break;
                     }
                 }
-                else
-                {
-                    // no more old changes, just add adjusted new change
-                    var adjustedNewChange = new TextChangeRange(new TextSpan(newChange.Span.Start - oldDelta, newChange.Span.Length), newChange.NewLength);
-                    AddRange(list, adjustedNewChange);
-                    newIndex++;
-                    goto nextNewChange;
-                }
             }
-            else
+
+            // there may be remaining old changes or remaining new changes (not both)
+            Debug.Assert(oldIndex == oldChanges.Length || newIndex == newChanges.Length);
+
+            while (oldIndex < oldChanges.Length)
             {
-                // no more new changes, just add remaining old changes
-                while (oldIndex < oldChanges.Length)
-                {
-                    AddRange(list, oldChanges[oldIndex]);
-                    oldIndex++;
-                }
+                add(oldChange); // no need to adjust oldDelta at this point
```

Comment:
why?  this comment needs explanation explaining why this would be the case.  (i'm not saying it's wrong, i'm just saying that the comment isn't evident).

Reply:
and might it not just be clearer and simpler to just do the addAndAdjustOldDelta so there is only one way we add tehse?  yes, there will be some additional adds.  but it's super unlikely that could ever actually affect perf here.

Reply:
> and might it not just be clearer and simpler to just do the addAndAdjustOldDelta so there is only one way we add tehse?

I decided to go with this because it ends up creating a quite pleasing uniformity across the method. Now `add` only gets called from `addAndAdjustOldDelta`, or from `addAdjustedNewChange`.